### PR TITLE
Add new parameters to snapshot restore to rename the restored aliases…

### LINKED ISSUE #15632: [Feature Request] Snapshot restore should allow renaming aliases
### Is your feature request related to a problem? Please describe

Snapshot restore takes a pair of arguments to allow renaming indexes into new names. The common example use case to restore snapshots into the same cluster as temporary separate indexes for research and debugging. However, this is problematic if this indexes has aliases to be restored. I am not sure if the indexes will be linked to the existing aliases or this is an error, but either case would problematic, since this either breaks desired isolation or fails.

### Describe the solution you'd like

Snapshot restore should take a new pair of arguments to allow renaming aliases into new names to match the existing pair of arguments for index renaming. For backend compatibility, these arguments should be optional.

### Related component

Storage:Snapshots

### Describe alternatives you've considered

In some cases, it may be possible to restore any necessary aliases after the snapshot restore is completed. It is also possible to restore into another cluster. The drawbacks of these two options is obvious, however.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mispencer**
### Description
Add new parameters to snapshot restore to rename the restored aliases similar to how you can currently rename the restored indexes.

### Related Issues
Resolves #15632

### Check List
- [x] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable. - [PR 615](https://github.com/opensearch-project/opensearch-api-specification/pull/615)
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable. - [PR 8544](https://github.com/opensearch-project/documentation-website/pull/8544)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
Please don't forget https://github.com/opensearch-project/opensearch-api-specification. See https://github.com/opensearch-project/opensearch-api-specification/pull/611 for recent changes in that space.

**Comment by mispencer** [comment]
I intended to wait until this PR has been reviewed, but ok. I have created related PRs in the doc and spec project.

**Comment by mispencer** [comment]
I assume the Changeling Verifier check fails is not meaningful. I also assume that the codecov/patch failure is not fatal.

**Comment by dbwiddis** [comment]
> I assume the Changeling Verifier check fails is not meaningful. I also assume that the codecov/patch failure is not fatal.

You can see which checks are required over on the right.  A changelog entry is required, while the coverage is advisory -- mostly a prompt for us to look more closely at whether tests were missed.

If you look at the change log failure, you can see that there's no backport tag on this PR, but the Change Log entry is listed for 2.x.   So if this is intended to be backported (as the change log suggests) we need to add the backport tag.

File: server/src/main/java/org/opensearch/action/admin/cluster/snapshots/restore/RestoreSnapshotRequest.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/action/admin/cluster/snapshots/restore/RestoreSnapshotRequest.java:174]: These version checks imply it will not be in 2.18.0, but you added the changelog in 2.x section implying they will.  Can you clarify?

Comment (mispencer) [reply server/src/main/java/org/opensearch/action/admin/cluster/snapshots/restore/RestoreSnapshotRequest.java:174]: I assume it can/should be in 2.18, but using V_2_18_0 caused the :qa:mixed-cluster:v2.18.0#mixedClusterTest to fail. You can see this failure in [one of the earlier Gradle checks](https://build.ci.opensearch.org/job/gradle-check/49318/consoleFull). I assume because this change is not back-ported into 2.18 yet. Back porting it I would assume would require this PR to be merged first. But merging this PR I assume would require that failing tests to pass. I am not sure how to resolve this catch-22.

Comment (dbwiddis) [reply server/src/main/java/org/opensearch/action/admin/cluster/snapshots/restore/RestoreSnapshotRequest.java:174]: Looks like previously it's handled by putting Version.CURRENT in the initial `main` PR, then once it's backported, updating the version in the backport and forward-porting:
https://github.com/opensearch-project/OpenSearch/commit/7c9c01d8831f57b853647bebebd8d91802186778

Comment (mispencer) [reply server/src/main/java/org/opensearch/action/admin/cluster/snapshots/restore/RestoreSnapshotRequest.java]: Ok, it has been updated with `Version.CURRENT`.

File: server/src/main/java/org/opensearch/snapshots/RestoreService.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/snapshots/RestoreService.java:496]: Some thoughts here.

1. the regex pattern will be compiled every iteration through this loop.  It's better to compile the pattern outside the loop once (`Pattern pattern = Pattern.compile(request.renameAliasPattern());`) and then inside the loop do `pattern.matcher(aliasName).replaceAll(request.renameAliasReplacement)`
2. related to the above, `replace()` is preferred to `replaceAll` if the match is not a regex.  Is there any way to test that, or provide another possible configuration?  This may not be that important depending on how often this is executed.
3. will we always want to `replaceAll()` vs. `replaceFirst()`?

File: server/src/main/java/org/opensearch/snapshots/RestoreService.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/snapshots/RestoreService.java:491]: This null check can be done outside the loop

File: server/src/main/java/org/opensearch/snapshots/RestoreService.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/snapshots/RestoreService.java:546]: Given we're repeating this logic from above, it might be best refactored into its own private method.

File: server/src/test/java/org/opensearch/snapshots/RestoreServiceIntegTests.java
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/snapshots/RestoreServiceIntegTests.java:41]: A 5-deep nested loop iterating over true/false?  I'm not sure that's better. 😄 

You could iterate an integer and use bitshifting and bitmasking to generate all the combinations?  Which brings up the question, there are 16 possible combinations but not all are used. Which ones aren't?  Might be worth a short comment.

File: server/src/test/java/org/opensearch/snapshots/RestoreServiceIntegTests.java
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/snapshots/RestoreServiceIntegTests.java:90]: I'm not fond of using underscores like this.  Are they really needed?

File: server/src/test/java/org/opensearch/snapshots/RestoreServiceIntegTests.java
Comment (mispencer) [reply server/src/test/java/org/opensearch/snapshots/RestoreServiceIntegTests.java:41]: I was actually thinking something like parameterized tests, but that doesn't seems to exists in JUnit 3.

File: server/src/main/java/org/opensearch/snapshots/RestoreService.java
Comment (mispencer) [reply server/src/main/java/org/opensearch/snapshots/RestoreService.java]: 1. Okay, done.
2. I don't see any trivial way to do that, and an index would have to have a LOT of aliases for this performance to matter.
3. This parameter pair is patterned off the existing rename_pattern/rename_replacement parameter pair which uses replaceAll. Though unlikely, there may be some use cases were replacing all is required, and I don't see any harm in using replace all.

File: server/src/main/java/org/opensearch/snapshots/RestoreService.java
Comment (mispencer) [reply server/src/main/java/org/opensearch/snapshots/RestoreService.java]: Okay, done.

File: server/src/main/java/org/opensearch/snapshots/RestoreService.java
Comment (mispencer) [reply server/src/main/java/org/opensearch/snapshots/RestoreService.java]: Okay, done.

File: server/src/test/java/org/opensearch/snapshots/RestoreServiceIntegTests.java
Comment (mispencer) [reply server/src/test/java/org/opensearch/snapshots/RestoreServiceIntegTests.java]: I think I just did this so it wouldn't be treated as a test, but I don't think the underscores is required. I have removed them.

Comment (dbwiddis) [reply server/src/test/java/org/opensearch/snapshots/RestoreServiceIntegTests.java]: I think since it has arguments it can't run as a test anyway.

File: server/src/main/java/org/opensearch/snapshots/RestoreService.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/snapshots/RestoreService.java]: 1. Awesome.
2. Agreed not worth the effort.
3. I'm fine with it if it matches the existing other parameter.

File: server/src/main/java/org/opensearch/snapshots/RestoreService.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/snapshots/RestoreService.java:670]: I think you can improve on this slightly without the extra variables.

In the case this conditional fails, you're still iterating the map just to add the keys to another set to return.

How about reversing this conditional (checking if either is null) and if so just return the keyset directly.

Then if you're continuing on you know the rename pattern is not null and you can apply the rest of the logic.

**Review by dbwiddis — COMMENTED** [review]
Looking good. One more performance improvement here

**Comment by dbwiddis** [comment]
@yigithub @jed326 @neetikasinghal Could one of you give this PR a once over as well?

**Review by dbwiddis — APPROVED** [review]
LGTM.  Would like someone familiar with the Snapshot code to approve this before merging.

File: server/src/test/java/org/opensearch/snapshots/RestoreServiceIntegTests.java
Comment (jed326) [reply server/src/test/java/org/opensearch/snapshots/RestoreServiceIntegTests.java:41]: We should be on Unit 4: https://github.com/opensearch-project/OpenSearch/blob/0f7d572da1d2dec4c2b83c6ae2c9e47ce38edd00/buildSrc/version.properties#L61

And there are other tests that are parameterized:
https://github.com/opensearch-project/OpenSearch/blob/0f7d572da1d2dec4c2b83c6ae2c9e47ce38edd00/server/src/internalClusterTest/java/org/opensearch/search/scroll/SearchScrollWithFailingNodesIT.java#L65-L71

**Comment by jed326** [comment]
Thanks @mispencer! Left one comment. The changes look good to me but I haven't been very hands on on Snapshots lately, would be great to get someone more involved like @sachinpkale or @ashking94 to give it another review.

File: server/src/test/java/org/opensearch/snapshots/RestoreServiceIntegTests.java
Comment (mispencer) [reply server/src/test/java/org/opensearch/snapshots/RestoreServiceIntegTests.java]: Okay - I have adjusted this.

**Comment by mispencer** [comment]
@dbwiddis - are your conversations resolved?

**Comment by dbwiddis** [comment]
> @dbwiddis - are your conversations resolved?

Yes but per the previous comment hoping @ashking94 or @sachinpkale can review. I'll merge tomorrow otherwise.

**Review by dbwiddis — APPROVED** [review]

**Review by dbwiddis — APPROVED** [review]

**Comment by mispencer** [comment]
@dbwiddis - do you intend to make a PR to update the versions in the [RestoreSnapshotRequest.java](https://github.com/opensearch-project/OpenSearch/pull/16292/files#diff-044b26faac340ba6801867e1c8a943fa978cc797c6d3a354ddc6f30f64702030R169), or do you want me to do that?

**Comment by dbwiddis** [comment]
> @dbwiddis - do you intend to make a PR to update the versions in the [RestoreSnapshotRequest.java](https://github.com/opensearch-project/OpenSearch/pull/16292/files#diff-044b26faac340ba6801867e1c8a943fa978cc797c6d3a354ddc6f30f64702030R169), or do you want me to do that?

@mispencer Please track this and follow it through to completion.  I (and most maintainers) are juggling multiple workstreams.

**Pull request merged**
